#Basic hands-on with Tensorflow:

TensorFlow is open-source end-to-end ML lib for preprocessing data,modelling data and serving models

Rather than building ML and DL from scratch,we use lib such as TensorFlow,it contains many most common ML functions you'll want to use,

Tensor is vast,but main premise is simple:turning the data into Tensors(numbers) and build ML.

The main difference in Numpy and Tensors arrays(also an n-d arrays of numbers)is that tensors can used on GPU(graphical processing units) and TPU(tensor processing units)

The first thing is import Tensorflow under the common alias tf.

In [ ]:
import tensorflow as tf
print(tf.__version__)#Finding  the version

2.17.1


Tensorflow has modules built-in(such as tf.io and tf.data)which are able to read your data sources and automatically convert them into tensors(numerics)and then later on ,neural netwrks models will process these for us.

constant() is used to create a Tensor from tensor like objects like List.

A scalar is knows as rank 0 tensor.because it has no dimensions(it's just number)

In [ ]:
#create a scalar (rank 0 tensor)
scalar=tf.constant(7)
scalar

<tf.Tensor: shape=(), dtype=int32, numpy=7>

In [ ]:
scalar.ndim#check for dimensions of tensor(ndim means number of dimensions)

0

In [ ]:
#vector as more than 0 dimensions
vector=tf.constant([10,10])
vector

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([10, 10], dtype=int32)>

In [ ]:
vector.ndim

1

In [ ]:
matrix=tf.constant([[10,7],[7,10]])
matrix


<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 7, 10]], dtype=int32)>

In [ ]:
matrix.ndim

2

In [ ]:

another_matrix=tf.constant([
   [[1,2,3],[5,6,4]], [[1,2,3],[4,5,6]]
    ],dtype=tf.float16)
another_matrix

<tf.Tensor: shape=(2, 2, 3), dtype=float16, numpy=
array([[[1., 2., 3.],
        [5., 6., 4.]],

       [[1., 2., 3.],
        [4., 5., 6.]]], dtype=float16)>

In [ ]:
another_matrix.ndim

3

dimensions 3,..this know as a rank 3 tensor.
For example, you might turn a series of images into tensors with shape (224, 224, 3, 32), where:
* 224, 224 (the first 2 dimensions) are the height and width of the images in pixels.
* 3 is the number of colour channels of the image (red, green blue).
* 32 is the batch size (the number of images a neural network sees at any one time).

All of the above variables we've created are actually tensors. But you may also hear them referred to as their different names (the ones we gave them):
* **scalar**: a single number.
* **vector**: a number with direction (e.g. wind speed with direction).
* **matrix**: a 2-dimensional array of numbers.
* **tensor**: an n-dimensional arrary of numbers (where n can be any number, a 0-dimension tensor is a scalar, a 1-dimension tensor is a vector).
![difference between scalar, vector, matrix, tensor](https://raw.githubusercontent.com/mrdbourke/tensorflow-deep-learning/main/images/00-scalar-vector-matrix-tensor.png)

#creating Tensor with tf.variable()
the difference between tf.variable() and tf.constant(),is tf.constant() are immutable (can't be changes,only be used to create new tensor),where as tensor created with tf.variable() are mutable.


In [ ]:
#creating a constant() and variable() with same
changeable_tensor_var=tf.Variable([10,6])
unchangeable_tensor_var=tf.constant([10,6])
changeable_tensor_var,unchangeable_tensor_var

(<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([10,  6], dtype=int32)>,
 <tf.Tensor: shape=(2,), dtype=int32, numpy=array([10,  6], dtype=int32)>)

In [ ]:
#using try and except method to handle the error.To change the value we need to use assign() method
#variable() is mutable as the value is changed
try:
  changeable_tensor_var[0].assign(3)
except Exception as e:
  print(e)
changeable_tensor_var

<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([3, 6], dtype=int32)>

In [ ]:
#using try and except method to handle the error.We're trying to change the constant()
#which cause the error as it's immutable
try:
  unchangeable_tensor_var[1].assign(3)
except Exception as e:
  print(e)


'tensorflow.python.framework.ops.EagerTensor' object has no attribute 'assign'


#Creating random tensors
random tensors are tensors of some abitrary size which contain random numbers.

purpose of random tensors is that neural networks use to intialize their weights(patterns)that they're trying to learn in the data.

The process of neural network learning often involves taking a random n-d array of numbers and refining them unitl they represent some kind of pattern

In [ ]:
#creating random tensors
va=tf.random.Generator.from_seed(42)
va1=va.normal(shape=(2,1))
va1

<tf.Tensor: shape=(2, 1), dtype=float32, numpy=
array([[-0.7565803 ],
       [-0.06854702]], dtype=float32)>

In [ ]:
#creating a 2 random(but the different)tensors
va2=tf.random.Generator.from_seed(42)
va2=va2.normal(shape=(2,1))
va3=tf.random.Generator.from_seed(1)
va3=va3.normal(shape=(2,1))
va2,va3

(<tf.Tensor: shape=(2, 1), dtype=float32, numpy=
 array([[-0.7565803 ],
        [-0.06854702]], dtype=float32)>,
 <tf.Tensor: shape=(2, 1), dtype=float32, numpy=
 array([[ 0.43842274],
        [-0.53439844]], dtype=float32)>)

In [ ]:
#creating a 2 random(but the same)tensors
va2=tf.random.Generator.from_seed(42)#set the seed for reproduciability
va2=va2.normal(shape=(2,1))#creating tensors from normal distrubtion
va3=tf.random.Generator.from_seed(42)
va3=va3.normal(shape=(2,3))
va2,va3,va2==va3

(<tf.Tensor: shape=(2, 1), dtype=float32, numpy=
 array([[-0.7565803 ],
        [-0.06854702]], dtype=float32)>,
 <tf.Tensor: shape=(2, 3), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702,  0.07595026],
        [-1.2573844 , -0.23193763, -1.8107855 ]], dtype=float32)>,
 <tf.Tensor: shape=(2, 3), dtype=bool, numpy=
 array([[ True, False, False],
        [False, False, False]])>)

In [ ]:
#shuffling the tensors using shuffle()
s=tf.random.shuffle([[1,2],
                     [3,5],[5,5]])
s

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[1, 2],
       [5, 5],
       [3, 5]], dtype=int32)>

In [66]:
#setting the shuffel order same by using seed()
tf.random.set_seed(42)#global random seed()
s1=tf.random.shuffle([[1,2],
                     [3,5],[6,5]],seed=42)#set the seed as global random see()
s1

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[1, 2],
       [3, 5],
       [6, 5]], dtype=int32)>

#creating a tensor for all value 1 using tf.ones()
and creating a tensors for all value 0 using tf.zeros()

In [71]:
t=tf.ones(shape=(3,3))
z=tf.zeros(shape=(3,3))
t,z

(<tf.Tensor: shape=(3, 3), dtype=float32, numpy=
 array([[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]], dtype=float32)>,
 <tf.Tensor: shape=(3, 3), dtype=float32, numpy=
 array([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]], dtype=float32)>)

#Matrix Multiplication
TensorFLow implemetns matrix multiplication functionality in the tf.matmul() method and other way  with python opertor "@".

the main 2 rules in matrix multiplications is:

1.The number of columns in the first matrix must be equal to the number of rows in the second matrix

2.The resulting matrix, known as the matrix product, has the number of rows of the first and the number of columns of the second matrix.

In [74]:
a=tf.random.Generator.from_seed(42)
a=a.normal(shape=(3,3))
a

<tf.Tensor: shape=(3, 3), dtype=float32, numpy=
array([[-0.7565803 , -0.06854702,  0.07595026],
       [-1.2573844 , -0.23193763, -1.8107855 ],
       [ 0.09988727, -0.50998646, -0.7535805 ]], dtype=float32)>

In [75]:
#using tensorflow built-in funtion matmul() to multiply matrix
tf.matmul(a,a)

<tf.Tensor: shape=(3, 3), dtype=float32, numpy=
array([[0.66619015, 0.02902636, 0.00942684],
       [1.0620725 , 1.0634612 , 1.6890633 ],
       [0.49040318, 0.49575394, 1.4989462 ]], dtype=float32)>

In [76]:
#using python operator "@" for matrix multiplcation
a@a

<tf.Tensor: shape=(3, 3), dtype=float32, numpy=
array([[0.66619015, 0.02902636, 0.00942684],
       [1.0620725 , 1.0634612 , 1.6890633 ],
       [0.49040318, 0.49575394, 1.4989462 ]], dtype=float32)>

#reshaping the matrix ,so it match the inner dimensions
 ex:if you have a=(2,3) and b=(5,3) shape of matix,you need to reshape the matrix b to become the b=(3,5) to match the inner dimensions(column of 1st matix and row of 2nd matrix)


In [87]:

a=tf.random.Generator.from_seed(42)
a=a.normal(shape=(2,3))
b=tf.random.Generator.from_seed(42)
b=b.normal(shape=(5,3))
a,b

(<tf.Tensor: shape=(2, 3), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702,  0.07595026],
        [-1.2573844 , -0.23193763, -1.8107855 ]], dtype=float32)>,
 <tf.Tensor: shape=(5, 3), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702,  0.07595026],
        [-1.2573844 , -0.23193763, -1.8107855 ],
        [ 0.09988727, -0.50998646, -0.7535805 ],
        [-0.57166284,  0.1480774 , -0.23362993],
        [-0.3522796 ,  0.40621263, -1.0523509 ]], dtype=float32)>)

In [88]:
#using reshape()-allows in tensor to reshape into defined shape
b=tf.reshape(b,shape=(3,5))
b,a

(<tf.Tensor: shape=(3, 5), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702,  0.07595026, -1.2573844 , -0.23193763],
        [-1.8107855 ,  0.09988727, -0.50998646, -0.7535805 , -0.57166284],
        [ 0.1480774 , -0.23362993, -0.3522796 ,  0.40621263, -1.0523509 ]],
       dtype=float32)>,
 <tf.Tensor: shape=(2, 3), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702,  0.07595026],
        [-1.2573844 , -0.23193763, -1.8107855 ]], dtype=float32)>)

In [92]:
#now we can do matmul() on a and b matrix
c=tf.matmul(a,b)
c

<tf.Tensor: shape=(2, 5), dtype=float32, numpy=
array([[ 0.7077842 ,  0.0272701 , -0.04926015,  1.0338199 ,  0.13473892],
       [ 1.1031651 ,  0.48607603,  0.6606892 ,  1.0202353 ,  2.3298066 ]],
      dtype=float32)>

In [96]:
#using transpose() method to match inner dimensions
d=tf.transpose(c)

d,tf.matmul(d,a)

(<tf.Tensor: shape=(5, 2), dtype=float32, numpy=
 array([[ 0.7077842 ,  1.1031651 ],
        [ 0.0272701 ,  0.48607603],
        [-0.04926015,  0.6606892 ],
        [ 1.0338199 ,  1.0202353 ],
        [ 0.13473892,  2.3298066 ]], dtype=float32)>,
 <tf.Tensor: shape=(5, 3), dtype=float32, numpy=
 array([[-1.9225984 , -0.304382  , -1.9438392 ],
        [-0.63181645, -0.11460861, -0.8781083 ],
        [-0.79347104, -0.14986205, -1.2001077 ],
        [-2.0649958 , -0.30749625, -1.7689085 ],
        [-3.0314033 , -0.5496058 , -4.2085466 ]], dtype=float32)>)

#typecasting in tensflow

In [97]:
A=tf.constant(5)#basically int datatype
A

<tf.Tensor: shape=(), dtype=int32, numpy=5>

In [99]:
A=tf.cast(A,dtype=tf.float32)#Cast() method is used to typecasting
A

<tf.Tensor: shape=(), dtype=float32, numpy=5.0>

#Getting the absolute value and findind the min,max,mean,sum,argmin,argmax

In [101]:
tf.abs(-12)#absolute value


<tf.Tensor: shape=(), dtype=int32, numpy=12>

aggression functions as syntax reduce_[action],
i.e,reduce_mean().
The functions are under the math module of tensor

In [105]:
#creating random values
import numpy as np
E=tf.constant(np.random.randint(0,50,25))#randint(low,high,size)
E

<tf.Tensor: shape=(25,), dtype=int64, numpy=
array([27, 23, 38, 21, 11, 39,  4, 44, 27, 39,  4, 10, 20, 35, 45, 48,  7,
       35, 35,  4, 42, 44,  9, 40,  6])>

In [106]:
tf.reduce_min(E)

<tf.Tensor: shape=(), dtype=int64, numpy=4>

In [107]:
tf.reduce_max(E)

<tf.Tensor: shape=(), dtype=int64, numpy=48>

In [108]:
tf.reduce_mean(E)

<tf.Tensor: shape=(), dtype=int64, numpy=26>

In [109]:
tf.reduce_sum(E)

<tf.Tensor: shape=(), dtype=int64, numpy=657>

In [110]:
tf.math.reduce_std(tf.cast(E,dtype=tf.float32))#standard deviation

<tf.Tensor: shape=(), dtype=float32, numpy=15.130156>

In [111]:
tf.math.reduce_variance(tf.cast(E,dtype=tf.float32))#variance

<tf.Tensor: shape=(), dtype=float32, numpy=228.9216>

In [112]:
tf.argmax(E)#return the index of max value


<tf.Tensor: shape=(), dtype=int64, numpy=15>

In [114]:
tf.argmin(E ) # return the index of min value

<tf.Tensor: shape=(), dtype=int64, numpy=6>

#Finding access to GPUs

If the a BELOW outputs an empty array (or nothing), it means you don't have access to a GPU (or at least TensorFlow can't find it).

In [2]:
import tensorflow as tf
print(tf.config.list_physical_devices('GPU'))

[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]


[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

the above is output if you got GPU access in Google Colab

In [3]:
!nvidia-smi


Thu Dec  5 10:19:30 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla T4                       Off | 00000000:00:04.0 Off |                    0 |
| N/A   33C    P8               9W /  70W |      3MiB / 15360MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+--



Thu Dec  5 10:19:30 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla T4                       Off | 00000000:00:04.0 Off |                    0 |
| N/A   33C    P8               9W /  70W |      3MiB / 15360MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+

Colab paid products - Cancel contracts here

if you have GPU  in your PC,above similar output is showed  
